storage-proofs/core/src/api_version.rs

```diff
@@ -0,0 +1,58 @@
+use std::fmt;
+use std::str::FromStr;
+
+use anyhow::{self, Result};
+use semver::Version;
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum ApiVersion {
+    V1_0_0,
+    V1_1_0,
+}
+
+impl fmt::Debug for ApiVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            ApiVersion::V1_0_0 => write!(f, "1.0.0"),
+            ApiVersion::V1_1_0 => write!(f, "1.1.0"),
+        }
+    }
+}
+
+impl fmt::Display for ApiVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            ApiVersion::V1_0_0 => write!(f, "1.0.0"),
+            ApiVersion::V1_1_0 => write!(f, "1.1.0"),
+        }
+    }
+}
+
+impl FromStr for ApiVersion {
+    type Err = anyhow::Error;
+    fn from_str(api_version_str: &str) -> Result<Self> {
+        let api_version = Version::parse(api_version_str)?;
+        match api_version.major {
+            1 => match api_version.minor {
+                0 => match api_version.patch {
+                    0 => Ok(ApiVersion::V1_0_0),
+                    _ => Err(anyhow::format_err!(
+                        "Could not parse API Version from string (patch)"
+                    )),
+                },
+                1 => match api_version.patch {
+                    0 => Ok(ApiVersion::V1_1_0),
+                    _ => Err(anyhow::format_err!(
+                        "Could not parse API Version from string (patch)"
+                    )),
+                },
+                _ => Err(anyhow::format_err!(
+                    "Could not parse API Version from string (minor)"
+                )),
+            },
+            _ => Err(anyhow::format_err!(
+                "Could not parse API Version from string (major)"
+            )),
+        }
```

Comment:
For the simple 2 case distinction, this makes sense.  I don't think this simplifies much for the general cases though (i.e. when we're actually using a range of versions across semver), in which case it needs to be broken out further into something like what's already there.

Reply:
Updated (minus the syntax error)

Reply:
You can match ranges in patterns too. I think if you order the clauses right, you *might* be able to use that fact to keep it relatively simple and legible. Maybe it will break down sometimes too, but it might be worth at least trying to isolate the complexity so remains mostly clear.

https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#matching-ranges-of-values-with-

Reply:
Just some more thoughts — nothing concrete to do now. Are you planning to use the semver matching facilities? In that case, the more complicated ranges could be encoded directly.

Example:
```
match api_version {
    v if VersionReq::parse("^1.2.3").matches(v) =>  todo!(),
    v if VersionReq::parse("^2.0").matches(v) => todo!(),
    v if VersionReq::parse("^1.0.0").matches(v) => todo!(),
}
```

I didn't think through the actual range matching in detail; this was just to indicate the form. Depending on what kinds of things come up, it might make sense to replace these explicit `matches` calls with custom predicates that can live in one place and capture the intended partitioning of versions as it evolves. If the same partitions are used in multiple decisions and become at all complicated, having a single source of truth to audit, update, and keep in sync between usages may be useful.